torch/lib/THD/base/data_channels/DataChannelGloo.hpp

```diff
@@ -0,0 +1,319 @@
+      std::make_shared<::gloo::AllreduceRing<T>>(
+        *context, {reinterpret_cast<T*>(input_buffer->data())},
+        count, THDToGlooReduceOp<T>(op)),
+      input_buffer,
+      input_buffer // we get the result in same buffer
+    );
+  }
+};
+
+template<typename T>
+struct algorithm_spec<DataOperation::BROADCAST, T> {
+  static GlooCache::key_type key(
+    THDGroup group_id, std::size_t input_bytes,
+    std::size_t count, rank_type src_rank // unused
+  ) {
+    return std::make_tuple(DataOperation::BROADCAST, group_id,
+                           input_bytes, input_bytes, 0);
+  }
+
+  static GlooCache::value_type create(
+    const DataChannel::Group& group,
+    std::size_t input_bytes, std::size_t count, rank_type src_rank
+  ) {
+    auto& cache = GlooCache::get();
+    auto context = cache.createContext(group);
+    auto input_buffer = cache.createBuffer(input_bytes);
+    return std::make_tuple(
+      context,
+      std::make_shared<::gloo::BroadcastOneToAll<T>>(
+        *context, {reinterpret_cast<T*>(input_buffer->data()}, count, src_rank
+      ),
+      input_buffer,
+      input_buffer // we get the result in same buffer
+    );
+  }
+};
+
+template<typename T> // unused
+struct algorithm_spec<DataOperation::BARRIER, T> {
+  static GlooCache::key_type key(THDGroup group_id) {
+    return std::make_tuple(DataOperation::BARRIER, group_id, 0, 0, 0);
+  }
+
+  static GlooCache::value_type create(const DataChannel::Group& group) {
+    auto context = GlooCache::get().createContext(group);
+    return std::make_tuple(
+      context,
+      std::make_shared<::gloo::BarrierAllToAll>(*context),
+      nullptr,
+      nullptr
+    );
+  }
+};
+
+
+struct GlooCache {
+  using key_type = std::tuple<
+    DataOperation, THDGroup, std::size_t, std::size_t, THDReduceOp
+  >;
+  using value_type = std::tuple<
+    std::shared_ptr<context_type>,
+    std::shared_ptr<algorithm_type>,
+    std::shared_ptr<buffer_type>,
+    std::shared_ptr<buffer_type>
+  >;
+
+  using ret_type = std::tuple<
+    std::shared_ptr<algorithm_type>, // algorithm
+    void*,                           // input buffer (nullptr if not used)
+    void*                            // output buffer (nullptr if not used)
+  >;
+
+  GlooCache(GlooCache const&)       = delete;
+  void operator=(GlooCache const&)  = delete;
+
+  // singleton instance
+  static GlooCache& get() {
+    static GlooCache instance;
+    return instance;
+  }
+
+  // TODO: enforce setters to be the first thing called (maybe some kind of factory)
+  void setStoreGetter(std::function<PrefixStore ()> store_getter) {
+    _store_getter = store_getter;
+  }
+
+  void setRank(rank_type rank) {
+    _rank = rank;
+  }
+
+  std::make_shared<context_type> createContext(const DataChannel::Group& group) {
+    // For now we have to get store even we are not in group to increase static `id` prefix
+    // See `getStore()` function implementation in `DataChannelGloo`.
+    auto store = _store_getter();
```

Comment:
No!!!! Just create a new `PrefixStore` for every call to a `DataChannelGloo` method! Most of them will be discarded, but it doesn't matter because their creation is nearly free!